### PR TITLE
Update trace queries to better handle time fields

### LINKED ISSUE #888: Imprecise trace representation
**What happened**: When viewing an OpenTelemetry Trace stored in ClickHouse, the trace is displayed with wrong timing information

Pasted image 20240619151413
*Figure 1: The ClickHouse trace viewed in grafana*

Pasted image 20240619151429
*Figure 2: The same trace stored and queried from Tempo viewed in Grafana*

The durations of the spans are correct, however the starting positions are not. The tempo trace is correct here.

**What you expected to happen**: I expect the ClickHouse trace representation to be the correct one.

**How to reproduce it (as minimally and precisely as possible)**: Setup any tracing solution and configure an OpenTelemetry collector to export the data at the same time to the ClickHouse database AND a Tempo Backend. View the same trace in both systems. Use the ClickHouse exporter contained in the contrib version of the collector.

**Screenshots**

**Anything else we need to know?**: This seems to relate to https://github.com/grafana/clickhouse-datasource/issues/771

This is the default query produced for viewing a trace:

```sql
WITH 'd469af085bf7524ba4dd5eb2814a6918' as trace_id, (SELECT min(Start)  
                                                      FROM "otel"."otel_traces_trace_id_ts"  
                                                      WHERE TraceId = trace_id) as trace_start, (SELECT max(End) + 1  
                                                                                                 FROM "otel"."otel_traces_trace_id_ts"  
                                                                                                 WHERE TraceId = trace_id) as trace_end  
SELECT "TraceId"                                                                                           as traceID,  
       "SpanId"                                                                                            as spanID,  
       "ParentSpanId"                                                                                      as parentSpanID,  
       "ServiceName"                                                                                       as serviceName,  
       "SpanName"                                                                                          as operationName,  
       "Timestamp"                                                                                         as startTime,  
       multiply("Duration", 0.000001)                                                                      as duration,  
       arrayMap(key -> map('key', key, 'value', "SpanAttributes"[key]), mapKeys("SpanAttributes"))         as tags,  
       arrayMap(key -> map('key', key, 'value', "ResourceAttributes"[key]),  
                mapKeys("ResourceAttributes"))                                                             as serviceTags  
FROM "otel"."otel_traces"  
WHERE traceID = trace_id  
  AND startTime >= trace_start  
  AND startTime <= trace_end  
LIMIT 1000
```
If it is adapted to 

```sql
WITH 'd469af085bf7524ba4dd5eb2814a6918' as trace_id, (SELECT min(Start)  
                                                      FROM "otel"."otel_traces_trace_id_ts"  
                                                      WHERE TraceId = trace_id) as trace_start, (SELECT max(End) + 1  
                                                                                                 FROM "otel"."otel_traces_trace_id_ts"  
                                                                                                 WHERE TraceId = trace_id) as trace_end  
SELECT "TraceId"                                                                                   as traceID,  
       "SpanId"                                                                                    as spanID,  
       "ParentSpanId"                                                                              as parentSpanID,  
       "ServiceName"                                                                               as serviceName,  
       "SpanName"                                                                                  as operationName,  
       "Timestamp"                                                                                 as startTimeNormal,  
       toUnixTimestamp64Nano(Timestamp) / 1000000                                                  as startTime,  
       multiply("Duration", 0.000001)                                                              as duration,  
       arrayMap(key -> map('key', key, 'value', "SpanAttributes"[key]), mapKeys("SpanAttributes")) as tags,  
       arrayMap(key -> map('key', key, 'value', "ResourceAttributes"[key]),  
                mapKeys("ResourceAttributes"))                                                     as serviceTags  
FROM "otel"."otel_traces"  
WHERE traceID = trace_id  
  AND startTimeNormal >= trace_start  
  AND startTimeNormal <= trace_end  
LIMIT 1000
```
mainly with `toUnixTimestamp64Nano(Timestamp) / 1000000 as startTime` will result in the correct trace representation:

Pasted image 20240619151547
*Figure 3: Trace view from ClickHouse with adapted query, now displayed correctly.*

**This is just a workaround / solution suggestion, if there are better ways to provide the same result, please let me know**

**Environment**:

- Grafana version: 11
- Plugin version: 4.1.0
- OS Grafana is installed on: Docker or Kubernetes
- User OS & Browser: Linux / Windows & Brave Browser
- Others:

### PR DESCRIPTION
**Pull request by aangelisc**
Traces are currently displayed with the wrong timing due to discrepancies between how the trace visualisation handles times and how the Javascript `Date` type functions (it doesn't correctly handle sub-millisecond precision).

Fixes #888 by converting the time field to nanoseconds and then adjusting it to milliseconds.

### CODE REVIEW

File: src/data/sqlGenerator.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import otel from 'otel';
  * Generates a SQL string for the given QueryBuilderOptions
  */
 export const generateSql = (options: QueryBuilderOptions): string => {
-  const hasTraceIdFilter = options.meta?.isTraceIdMode && options.meta?.traceId
+  const hasTraceIdFilter = options.meta?.isTraceIdMode && options.meta?.traceId !== undefined;
```
Comment (SpencerTorres) [inline src/data/sqlGenerator.ts:8]: What if `traceId` is empty string?

**Review by SpencerTorres — CHANGES_REQUESTED** [review]
Overall looks good, just the one question. Thanks for submitting this!

File: src/data/sqlGenerator.ts (hunk 1 above)
Comment (aangelisc) [reply src/data/sqlGenerator.ts]: Good catch. I've reverted that change, my bad!

**Review by SpencerTorres — APPROVED** [review]

**Review by adamyeats — APPROVED** [review]

**Comment by rudebono** [comment]
Thank you ⸜(｡˃ ᵕ ˂ )⸝♡

**Pull request merged**